cephfs,mon: disallow non-empty data pool for "fs new" and "fs add_data_pool" cmd

rishabh-d-dave opened the pull request:
Instead of accepting non-empty data pool names exit with error code EINVAL and an appropriate error message.

Fixes: https://tracker.ceph.com/issues/70361
Fixes: https://tracker.ceph.com/issues/70362






<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

rishabh-d-dave committed:
cephfs,mon: don't allow non-empty data pool for "fs new" cmd

Instead of allow non-empty data pools for command "ceph fs new", abort
with error number EINVAL and an appropriate error message. This will
automatically have the same effect on "ceph fs volume create" command
too.

Fixes: https://tracker.ceph.com/issues/70361
Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
qa/cephfs: test that non-empty data pool is not allowed for volume...

creation

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
cephfs,mon: disallow non-empty data pools for "fs add_data_pool" cmd

Instead of allow non-empty data pools for "ceph fs add_data_pool"
command, abort with error number EINVAL and print an appropriate
message.

Fixes: https://tracker.ceph.com/issues/70362
Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave commented:
@batrick Let me know if I should proceed to add tests and making rest of the required modifications. cc @ceph/cephfs

batrick commented on src/mon/FSCommands.cc:
```diff
@@ -213,6 +213,16 @@ class FsNewHandler : public FileSystemCommandHandler
+    const pool_stat_t *data_pool_stat = mon->mgrstatmon()->get_pool_stat(data);
+    if (data_pool_stat) {
+      int64_t data_num_objects = data_pool_stat->stats.sum.num_objects;
+      if (!force && data_num_objects > 0) {
```
add `dout` that `--force` is overriding this check

rishabh-d-dave replied:
Done.

batrick commented on src/mon/FSCommands.cc:
```diff
@@ -1084,6 +1094,16 @@ class AddDataPoolHandler : public FileSystemCommandHandler
+    if (pool_stat) {
+      int64_t data_num_objects = pool_stat->stats.sum.num_objects;
+      if (data_num_objects > 0) {
+	ss << "pool '" << poolname  << "' already contains some objects. Use "
```
it should be possible to override this too

rishabh-d-dave replied:
Done.

rishabh-d-dave committed:
cephfs,mon: small optimization for "fs new" code

During check for metadata pool's emptiness in "ceph fs new" command's
code, it's pointless to get stat for the metdata pool and to get the
count for the objects in it when --force has been passed. --force
ensures that non-empty metadata pool will be used for creating new FS.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
cephfs,mon: tell user that --force can allow using non-empty metadata...

pool for creating new FS.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
qa/cephfs: test passing metadata pool to "fs new" with and...

without --force.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
qa/cephfs: tests passing data pool to "fs new" with and...

without --force.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
qa/cephfs: test "fs add_data_pool" for non-empty data pool with and...

without --force.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
mgr/vol: allow --force for 'volume create" command to pass non-empty...

data and metadata pools.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
qa/cephfs: test "volume create" command with --force

Test that "ceph fs volume create" command runs successfully when
non-empty data and metadata pool names are passed along with "--force".

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave commented:
The first 5 commits are from a different PR on which some of the code from this PR depends on. Once, that PR merges, I'll rebase this PR branch and those extra commits will be vanish from this PR's branch.

anthonyeleven commented on PendingReleaseNotes:
```diff
@@ -90,6 +90,11 @@
+* CephFS: Command "ceph fs volume create" now allows users to pass metadata
+  and data pool names that should be used for creating the volume. In case
+  the passed name is a name of a non-empty pool, the command will abort.
+
```
Don't I recall reviewing this change a couple of days ago in another PR?

Suggest:

* CephFS: The `ceph fs volume create` command now allows admins to supply metadata
   and data pool names to be used when creating the volume. If any of these pools
  exists and is non-empty pool, the command will abort.

rishabh-d-dave replied:
I guess you missed my  reply. There are 3 separate features in PR and therefore clubbing all of them into one release note is not a good idea. Plus, review only latter 2 features, there is a separate PR for former feature. For more details please check earlier replies.

mchangir commented on src/mon/FSCommands.cc:
```diff
@@ -208,8 +208,9 @@ class FsNewHandler : public FileSystemCommandHandler
 	if (metadata_num_objects > 0) {
-	  ss << "pool '" << metadata_name
-	     << "' already contains some objects. Use an empty pool instead.";
+	  ss << "pool '" << metadata_name << "' already contains some objects. "
+	        "Use an empty pool instead. Pass --force if you are sure that "
+		"you wish to continue";
```
rephrase as ...
```
Pass --force if you are sure that you wish to continue using the dirty pool.
```

mchangir commented on qa/tasks/cephfs/test_admin.py:
```diff
@@ -499,6 +499,76 @@ class TestFsNew(TestAdminCommands):
+        # that the meta pool is not empty and therefore abort with an error.
+        sleep(5)
+
+        try:
+            # actual test...
+            self.run_ceph_cmd(f'fs new {fs} {meta} {data} --force')
+
+            # being extra sure that volume wasn't created
+            output = self.get_ceph_cmd_stdout('fs ls').lower()
+            self.assertNotIn(fs, output)
```
I presume that the fs **should** get created with the `--force` option.
If so, then the assertion `self.assertNotIn(fs, output)` is not required here ?
The code **should** assert that the fs is indeed listed in the output.

mchangir commented on src/mon/FSCommands.cc:
```diff
@@ -214,6 +214,17 @@ class FsNewHandler : public FileSystemCommandHandler
+      if (data_pool_stat) {
+	int64_t data_num_objects = data_pool_stat->stats.sum.num_objects;
+	if (!force && data_num_objects > 0) {
+	  ss << "pool '" << data_name << "' already contains some objects. Use "
+		"an empty pool instead. Pass --force if you are sure that you "
+		" wish to continue";
```
rephrase as ...
```
Pass --force if you are sure that you wish to continue with the dirty pool.
```

mchangir commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -841,6 +876,42 @@ def test_with_nonempty_data_pool_name(self):
+                              retval=errno.EINVAL)
+
+            # being extra sure that volume wasn't created
+            output = self.get_ceph_cmd_stdout('fs ls').lower()
+            self.assertNotIn(v, output)
```
shouldn't the assertion be **assertIn()** instead ?

rishabh-d-dave commented on src/mon/FSCommands.cc:
```diff
@@ -208,8 +208,9 @@ class FsNewHandler : public FileSystemCommandHandler
-	  ss << "pool '" << metadata_name
-	     << "' already contains some objects. Use an empty pool instead.";
+	  ss << "pool '" << metadata_name << "' already contains some objects. "
+	        "Use an empty pool instead. Pass --force if you are sure that "
```
This PR doesn't add `--force`, it is already present. This commit only modifies the message to tell user about --force option.

rishabh-d-dave commented:
jenkins test all

rishabh-d-dave commented:
jenkins retest this

rishabh-d-dave commented:
jenkins retest this please

rishabh-d-dave commented:
Rebased.
